### DISCUSSION TITLE
Does Vercel functions support http request streaming?

### DISCUSSION THREAD

**Discussion by zizifn** [vercel/community#1037] [Help] [answered]
Now fetch in Chrome support streaming HTTP request.
https://developer.chrome.com/articles/fetch-streaming-requests/ 

When I tested this in nextjs, It's supported in my locally, but it does not support in Vercel serverless function.  So does HTTP request streaming is possible in Vercel serverless function?

Below is my API route,
``` typescript
import type { NextApiRequest, NextApiResponse } from "next";
import type { Readable } from "node:stream";

export const config = {
  api: {
    bodyParser: false,
  },
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  console.log("start");
  if (req.method === "POST") {
    let index = 0;
    res.writeHead(200);
    for await (const chunk of req) {
      console.log(`client send ${chunk.toString()}`);
      res.write(
        `server response ${index++}, after client send ${chunk.toString()}`
      );
    }
    res.end();
  } else {
    res.setHeader("Allow", "POST");
    res.status(405).end("Method Not Allowed");
  }
}

```

And below is my test script in `node.js`
```javascript
import {
  ReadableStream,
  TextDecoderStream,
  TextEncoderStream,
} from "node:stream/web";
import { fetch } from "undici";

try {
  const readableStream = new ReadableStream({
    index: 0,
    start(controller) {
      const interval = setInterval(() => {
        controller.enqueue(`client send ${this.index++}`);
      }, 500);

      setTimeout(() => {
        clearInterval(interval);
        controller.close();
      }, 10_000);
    },
  }).pipeThrough(new TextEncoderStream());

  const url = "https://*****/api/proxy2";
  // const url = "http://localhost:3000/api/proxy2";
  const resp = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "text/plain",
    },
    body: readableStream,
    duplex: "half",
  });

  for await (let chunk of resp.body.pipeThrough(new TextDecoderStream())) {
    console.log(chunk);
  }
} catch (err) {
  console.log(err);
}

```

**Comment by amyegan** [comment, marked as answer]
Yes, Vercel Functions support streaming: https://vercel.com/docs/concepts/functions/edge-functions/streaming

**Reply by zizifn** [reply]
Any plan for this or there is underlying technical issue?

**Reply by amyegan** [reply]
As it turns out, I was wrong! Sorry about that. [Streaming with Edge Functions](https://vercel.com/docs/concepts/functions/edge-functions/streaming#) is possible. Just found the new documentation mentioning [streaming](https://vercel.com/docs/functions/streaming) support for Serverless Functions, Edge Functions, and React Server Components in Next.js projects.

**Reply by zizifn** [reply]
The doc is about streaming response body. I am checking does Vercel is also support REQUEST body streaming.

**Reply by zizifn** [reply]
>Now fetch in Chrome support streaming HTTP request.
https://developer.chrome.com/articles/fetch-streaming-requests/

**Comment by isaiahtaylor** [comment]
FYI - I finally got streaming to work on serverless functions so I can stream OpenAI. See: https://twitter.com/isaiah_p_taylor/status/1638216571507331078

**Reply by weleo** [reply]
What's your solution? I'm not seeing any in the Twitter thread. Could you please share it with us?

**Reply by isaiahtaylor** [reply]
@weleo updated that thread. Apparently streaming support is only for the new `app` serverless functions.... but those have a module loading issue with langchain.

**Reply by zizifn** [reply]
I remember I saw someone twitter. Seems langchian support serverless, maybe I am wrong....

**Reply by alibey-10** [reply]
@weleo check this article [link](https://vercel.com/blog/gpt-3-app-next-js-vercel-edge-functions)

**Reply by zhuchaokn** [reply]
I am using the Flask framework in Python and encountered the same problem https://github.com/orgs/vercel/discussions/2473. 
I was able to reproduce this problem on my personal cloud server. I think the reason for this may be that the Vercel reverse proxy did not add the HTTP 1.1 identifier. For example, in the configuration under Nginx: 
```
proxy_http_version 1.1;
 proxy_set_header Connection ""; 
```
I added these two lines to the configuration of Nginx, and the problem was solved.

**Reply by leerob** [reply]
Vercel does not currently support streaming with the Python runtime.

**Comment by jasan-s** [comment]
So is it official that langchain python will not work on vercel if looking for streaming responses.

**Comment by kn** [comment]
Pinging about the original questions:
Is there plan to support streaming for request (not response)?

**Comment by alexvuka1** [comment]
@amyegan's answer does not answer the question. Could we please get an appropriate answer regarding streaming **request** bodies?
